### ISSUE TITLE
[RFC]: Model Parallelism with Single Worker using SPMD

### ISSUE THREAD

**Issue by lsy323** [vllm-project/vllm#18009]
## Goal
Propose to add an alternative to enable model parallelism with a single worker, partition model with GSPMD.

## Motivation

![Image](https://github.com/user-attachments/assets/872501e3-a1bf-45d2-a229-e401236e127b)

_Figure 1: (Left) Current vLLM multi-chip architecture (Right) Proposed single worker architecture_

The current vLLM utilizes multiple workers when a model is partitioned across several devices (**One device per worker**). In this setup, communication operations, such as reducing partial sums or gathering sharded data, are explicitly called within the model's implementation.

In contrast, the PyTorch on TPU stack (PyTorch/XLA and the XLA Compiler) offers an alternative parallelism programming model: Programmer implements models **as if it runs on a single device (No explicit communication op)**, and **provides hints to compiler** on how to partition the model. The compiler then automatically inserts the necessary collective communication operations and applies optimization passes to overlap communication and computation. Therefore, the model code is expected to have **global shape** even if the model is partitioned. The TPU compiler stack is heavily based on this programming model, which was proposed in the [GSPMD](https://arxiv.org/pdf/2105.04663) paper.

Importantly, this model still offers developers the flexibility to explicitly invoke communication operations, allowing them to override the compiler and define a partitioned program manually, similar to vLLM's current programming model.

At the runtime level, all devices within the same host can be managed by a single process, executing the same compiled program. In multi-host, each host has a process managing its local devices.

Here is an [example](https://github.com/lsy323/learning_machine/blob/lsiyuan/torch-xla-example-vllm/torchxla-for-vllm/ffn_spmd.py#L57-L60) to do Megatron style sharding on a FFN module by applying compiler hints using the SPMD API in PyTorch/XLA. For more details on SPMD x PyTorch/XLA, please check this [guide](https://github.com/pytorch/xla/blob/master/docs/source/perf/spmd_basic.md).

## Benefits

1. There are advanced optimizations that can only be enabled through the above compilation pass. Therefore enabling the above programming model in vLLM is essential to the multi-chip inference performance on TPUs.
2. With the new parallel programming model, model code is not coupled with the parallel configuration anymore. This provides vLLM developers with an alternative way to implement model parallelism. (Similar to the programming model in torchtitan, [example](https://github.com/pytorch/torchtitan/blob/508350b7d6f6ca32d48f414852feda5004a7720f/torchtitan/experiments/flux/parallelize_flux.py#L55))

## Combining single worker mode and multi-worker mode

Developers should be able to combine the single worker mode with the multi-worker mode. For example, when pipeline parallelism and tensor parallelism are enabled, each PP stage can have 1 worker, running the TP program on tp_size devices. Since PP is not supported on the TPU backend yet, this use case won’t be covered in this RFC.

### Use environment variable
We propose to use an environment variable to enable the single worker mode. This may be cleaner because we plan to use single-worker as the default distributed backend for TPU. At that time the new configuration flag is not needed anymore.

## Executor Configuration
In vLLM, different executors are used for single-process and multi-process execution environments. Single-process uses uni executor and multi-process uses mp or ray executor. The executor type is defined in this [field](https://github.com/vllm-project/vllm/blob/main/vllm/config.py#L1667-L1668) of the ParallelConfig, and it’s derived from the other parallel configurations (e.g. if TP size is 1, then use uni-process executor, if TP size > 1, then use multi-process executor) [code [ref](https://github.com/vllm-project/vllm/blob/main/vllm/config.py#L1753-L1819) for determine executor type].

we can extend the logic of determining distributed executor type to use a single process executor under single worker mode. Here is the code [change](https://github.com/lsy323/vllm/blob/afa57bc83607f398d92a2954ef8f012c07c62802/vllm/config.py#L1808-L1809) in the prototype.

## Runtime configuration
We need to configure runtime to control multiple devices within one process. (By default it’s 1 device / process) This can be done during [init_device](https://github.com/lsy323/vllm/blob/afa57bc83607f398d92a2954ef8f012c07c62802/vllm/v1/worker/tpu_worker.py#L266-L267) with only 1 line change for TPU

## Model initialization / Weight loading

![Image](https://github.com/user-attachments/assets/9601263b-f038-42d8-bccd-5b081e4c5630)
_Figure 2: Current model initialization and weight loading workflow for multi-chip setup_

![Image](https://github.com/user-attachments/assets/422435b2-f0ec-46d5-a1af-2dbe000a6d1e)
_Figure 3: Model initialization, weight-loading, and model partition for single worker mode_

### Overwrite parallelism configuration
Because we want the model to have global shapes, the model needs to be initialized the same way as (TP=1, EP=1, PP=1). We can tweak the configuration before model loading in [tpu_model_runner.py](https://github.com/vllm-project/vllm/blob/main/vllm/v1/worker/tpu_model_runner.py#L889)

### Meta device init, cpu weight loading, transfer to TPU during partitioning

Initializing the full model on CPU may cause host OOM if the model is very large (e.g. DeepSeek v3  - more than 1.2 TB).

To make the model initialization and weight loading more efficient and cleaner, avoid OOM during model init/weight-loading, the proposed workflow will be based on the following:

1. Initialize model on [Meta](https://pytorch.org/docs/stable/meta.html) device. Initializing models on TPU has the following drawbacks. Specifically, we need to avoid the following code
```
with target_device:
  model = _initialize_model(vllm_config)
```

  - Drawback 1: Use more memory than the size of the model weights, this leads to OOM even if the size of the model can fit on device. The reason is model initialization logic can have reshape operations, which is not an in-place operation on TPU, requiring additional memory

  - Drawback 2: Initializing the model on TPU would invoke graph compilation and execution, which will significantly slow down the model initialization process.

2. Weight-loading on CPU
If we load the weight on TPU, the drawback is similar to that of doing model initialization on TPU.
  - Drawback 1: Weight loading logic has reshape, view, transpose op, which may not be in-place ops on TPU.
  - Drawback 2: Weight loading on TPU triggers graph compilation and execution, making the weight-loading slow.

Here is the proposed model initialization, weight loading, host->device transfer workflow for TPU (Please see figure 3 for details):

1. Initialize full model on Meta device, which doesn’t allocate physical memory on CPU
2. Add a pre-hook for the load_weight function, which move the weight tensor to CPU
3. Call the load_weight function to load the weight on CPU
4. Add a post-hook to the load_weight to partition the weight (By applying compiler hint) and then transfer to device.

Proposed change:
Add a TPU loader under [vllm/model_executor/model_loader](https://github.com/vllm-project/vllm/tree/main/vllm/model_executor/model_loader).

## Model Partition

We propose to add a new file [vllm/distributed/tpu_distributed_utils.py](https://github.com/lsy323/vllm/blob/lsiyuan/spmd-merge/vllm/distributed/tpu_distributed_utils.py) to process the loaded model on CPU to:

1. Attach compiler hints for model partitioning 
2. Transfer shards to devices (Implicitly happens when the partitioning hint is applied to weight tensors)
3. Module swap: Replace original model layers with a new layer. For example, we will have the flexibility to replace the FFN module in the model with an optimized kernel in which communication and computation are overlapped.

To achieve this, for each parallel layer involving collective ops in vLLM (e.g. [RowParallelLinear](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/linear.py#L1126), [QKVParallelLinear](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/linear.py#L784)), we implement a function to apply the compiler hints to let the compiler know how the layer is partitioned ([example](https://github.com/lsy323/vllm/blob/afa57bc83607f398d92a2954ef8f012c07c62802/vllm/distributed/tpu_distributed_utils.py#L102-L107)). Then we recursively [process](https://github.com/lsy323/vllm/blob/afa57bc83607f398d92a2954ef8f012c07c62802/vllm/distributed/tpu_distributed_utils.py#L149-L168) the model based on the type of each submodule in the model.

### Why module swap is needed

![Image](https://github.com/user-attachments/assets/3fc09fef-77f8-42bc-815f-c28026ff9e3e)
_Figure 4: QKVParallelLinear with Tensor Parallelism_

![Image](https://github.com/user-attachments/assets/dc8b9f09-b91c-45c4-b610-850ecd3e88dd)
_Figure 5: Partition the full shape weight of QKVParallelLinear_

For QKVParallel Linear, the `q_proj`, `k_proj` and `v_proj` weights are partitioned along the `num_heads/kv_heads * head_dim` dimension (See figure 4 for details). To simplify the problem, let’s assume `num_kv_heads` is greater than and divisible by `tp_size` here.

However, since the model is initialized with full shape, the `qkv_proj` will be 1 large tensor of shape `[hidden_size, (num_heads + 2 * num_kv_heads) * head_dim]`. Plus when partitioning a tensor with SPMD, the tensor needs to be partitioned evenly along an axis. Partitioning the qkv combined weights will result in an inefficient computational graph (As shown in figure 5).

To make SPMD performant in vLLM, the `q/k/v_proj` weights need to be stored in 3 separate weight tensors separately, so that the weight partition can be done properly (Shown in figure 4). Therefore, we propose to do module swap to replace the QKVParallelLinear layer with a more SPMD friendly QKVParallelLinear implementation.

## KV Cache partition
The partition annotation can be applied in TPU model runner, when the kv cache tensors are initialized and binded to the model. Here is an example for TP ([example](https://github.com/lsy323/vllm/blob/afa57bc83607f398d92a2954ef8f012c07c62802/vllm/v1/worker/tpu_model_runner.py#L1327-L1328))

One edge case is when the `num_kv_heads` is smaller than `tp_size`, we need to ensure each partition has at least 1 kv_head. This handling needs to be done when the kv cache tensors are created (In the current flow, it’s handled in model code [ref](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/llama.py#L128)).

## Torch.compile integration
Currently, torch.compile is applied by a decorator in the model code ([example](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/llama.py#L290)). We need to manually apply the same torch.compile wrapper again in TPU model runner.

## Profiling run, warmup run, execute_model
No change is needed, since the input / output of the model is not changed.

## Scheduler and other components in Engine
No change, since all the changes are limited to TPU Worker, TPU Model Runner

### Feedback Period.

2 weeks

### CC List.

@WoosukKwon @robertgshaw2-redhat @simon-mo @NickLucche @tlrmchlsmth @alexm-redhat @mgoin @youkaichao @DarkLight1337 @yaochengji @yarongmu-google @bythew3i @vanbasten23 @Chenyaaang @fenghuizhang @richardsliu @yixinshi @miladm @bvrockwell 

### Any Other Things.

In this RFC, we introduce a new parallel programming model to vLLM, to accommodate the TPU compiler stack. We would like to hear the comments and feedback from the community on this programming model, in case it also benefits other vLLM developers and researchers!

## Acknowledgement

@yaochengji: Thank you for brainstorming and proposing ideas on the SPMD x vLLM integration and many general parallelism knowledge!

@hosseinsarshar: Thank you for working on a prototype of adopting SPMD in vLLM! Thank you for proposing to use `tpu_info` to workaround `xm.get_memory_info` is not supported under SPMD in torch_xla.

@qihqi: Thank you for many suggestions and ideas around the SPMD x vLLM integration. Also thank you for pointing out how to make QKVParallelLinear performant with SPMD!

Also thank you @mgoin @WoosukKwon @robertgshaw2-redhat @tlrmchlsmth @alexm-redhat @NickLucche @bythew3i @vanbasten23 @fenghuizhang for the early feedbacks in the offline review session!
 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by NickLucche** [comment]
Mega job here, looking forward to testing it out!
It would also be nice to distill a little of this RFC into a docs page for TPU-model parallelism.


One thing I found slightly confusing is this weight loading part:
```
2. Weight-loading on CPU
If we load the weight on TPU, the drawback is similar to that of doing model initialization on TPU.
```
isn't this related to loading on TPU? By `Weight-loading on CPU` I thought you meant actually loading the data (not just meta) on cpu, applying transforms, and then doing a .to(tpu).

**Comment by lsy323** [comment]
Hi @NickLucche, thank you for the comment! And sorry for the late reply!

> It would also be nice to distill a little of this RFC into a docs page for TPU-model parallelism.

Second this, once the RFC is agreed in the vLLM community, we can put it in a doc and continuously update it as we go.

> isn't this related to loading on TPU? By Weight-loading on CPU I thought you meant actually loading the data (not just meta) on cpu, applying transforms, and then doing a .to(tpu).

Sorry for the confusion, I should be more clear on this part. The weight loading here refers to loading HuggingFace weights to the `nn.Module` instance. This process (HF weights -> `nn.Module`) happens on CPU. Currently, the the `nn.Module` is on TPU, and HF weights are on CPU, and the weight loading is like `tpu_weights.data.copy_(cpu_weights)`.

After the HF weights are loaded into the CPU `nn.Module`, we do the weight partition and transfer to TPU devices.

**Comment by youkaichao** [comment]
the vllm neuron backend used to take this approach, but it does not really work well. the problem is you need to change the model files, like `llama.py` , which we never want to introduce hardware-specific code.

I would suggest trying to stick with the 1 device per process approach as long as you can, otherwise tpu code can be easily broken and will be less maintainable for the rest maintainers.

**Comment by lsy323** [comment]
Hi @youkaichao, thank you for the comment! 

> I would suggest trying to stick with the 1 device per process approach as long as you can, otherwise tpu code can be easily broken and will be less maintainable for the rest maintainers.

Agreed that using 1 process per device is more aligned with the vLLM design (or the overall PyTorch GPU paradigm) , however, the current state is the XLA compiler stack only works well, reaches best performance through the SPMD workflow.

> the vllm neuron backend used to take this approach, but it does not really work well. the problem is you need to change the model files, like `llama.py` , which we never want to introduce hardware-specific code.

For simple cases (like linear layers), code changes to the model implementation are not needed ([example](https://github.com/lsy323/vllm/blob/lsiyuan/spmd/vllm/distributed/tpu_distributed_utils.py#L111-L116)). During host->device transfer, the weights are partitioned onto difference devices, and that's enough for compiler to figure out we are doing Megatron sharding for TP.

For more complex cases (QKVParallelLinear), code changes to model code is needed, but we can do module swap after the model initialization ([example](https://github.com/lsy323/vllm/blob/lsiyuan/spmd/vllm/distributed/tpu_distributed_utils.py#L127-L132)). In this way, vLLM model code doesn't need to be changed, or accommodate to the SPMD workflow. Also, the necessity of doing module swap for `QKVParalleLinear` is explained in the `Why module swap is needed` section in the RFC.

Even though we haven't done comprehensive testing among a variety of models, I think the module swap will only applied to a limited number of layers (like `QKVParallelLinear`).

**Comment by ywang96** [comment]
@youkaichao @NickLucche Could we come out with a conclusion here? Thanks!
